### ISSUE TITLE
declaration-block-no-redundant-longhand-properties

### ISSUE THREAD

**Issue by evilebottnawi** [stylelint/stylelint#1735]
Ref: https://github.com/CSSLint/csslint/wiki/Require-shorthand-properties
Ref: https://github.com/CSSLint/csslint/issues/668

I don't know best name for rule :smile: 

Throw error if we use all 4 values, also throw errors if we use top and left, top and left and right.
Example:
```css
a {
  padding-top: 1px;
  padding-right: 2px;
  padding-bottom: 3px;
  padding-left: 4px; /* Throw error, need `padding: 1px 2px 3px 4px;` */
}
```

```css
a {
  padding-top: 5px;
  padding-right: 5px;
  padding-bottom: 5px; /* Throw error, need `padding: 1px 2px 3px;` */
}
```

```css
a {
  padding-top: 5px;
  padding-right: 5px; /* Throw error, need `padding: 1px 2px;` */
}
```
Also add option `minProperties`(maybe rename) to control how many properties throw errors.

**Comment by Arcanemagus** [comment]
Might want to create a label or milestone to track these "CSSLint compatibility" issues.

**Comment by davidtheclark** [comment]
Milestone sounds good. Just created one.

**Comment by jeddy3** [comment]
Lets only check for when *all four* longhand properties are used. Warning when there are only *three* or *two* longhand properties will lead to second guessing the author's intent e.g.

```css
a {
  margin: 10px;
}

/* the authors intent is to override everything *but* the top margin */
a.primary {
  margin-bottom: 0;
  margin-left: 0;
  margin-right: 0;
}
```

I assume that's why the CSSLint rule only checks for *all four* properties too.

**Comment by evilebottnawi** [comment]
@jeddy3 example without `top` and no throw errors, only if you have all 4 options, or (top and right and bottom), or (top and right)

**Comment by jeddy3** [comment]
> or (top and right and bottom), or (top and right)

I'm strugglingly to understand this bit. Can you explain it again? 

Doesn't:

```css
padding: 1px 2px;
```

Equal:

```
padding-top: 1px;
padding-bottom: 1px;
padding-left: 2px;
padding-right: 2px;
```

**Comment by evilebottnawi** [comment]
@jeddy3 
if i use
```
a {
  padding-top: 1px;
  padding-right: 2px;
  padding-bottom: 3px; 
}
```
I can short to `padding: 1px 2px 3px`, but only if i use `top`, `right` and `bottom`, if i have no `top` or `right` not throw error.

**Comment by davidtheclark** [comment]
I'll try to help clarify. I think the rule looks for `padding-[side]` declarations, and then if all of them are used suggests that you use the `padding` shorthand.

**Comment by davidtheclark** [comment]
> Also add option minProperties(maybe rename) to control how many properties throw errors.

I don't see this in the CSSLint rule so let's not consider clarifying and creating it essential for now.

**Comment by evilebottnawi** [comment]
@davidtheclark i thinkw e should throw errors on all variants:
```css
a {
  padding-top: 1px;
  padding-right: 2px;
  padding-bottom: 3px;
  padding-left: 4px; /* Throw error, need `padding: 1px 2px 3px 4px;` */
}
```

```css
a {
  padding-top: 5px;
  padding-right: 5px;
  padding-bottom: 5px; /* Throw error, need `padding: 1px 2px 3px;` */
}
```

```css
a {
  padding-top: 5px;
  padding-right: 5px; /* Throw error, need `padding: 1px 2px;` */
}
```
But `CSSLint` throw error only on 4 properties, `"minProperties": 4` allow throw error only on first example (need for csslint-compat).

**Comment by davidtheclark** [comment]
I disagree. A major problem with shorthands is that they actually *always set all the properties*, even if you only *want* to specify some. So `padding: 5px 0;` is *not* equivalent to `padding-top: 5px; padding-bottom: 5px` because it also sets left and right padding, which you may not want to do (probably don't). I think maybe the only way for the rule to avoid forcing you to change your code in ways that you actually don't want is to *only* error when all the shorthand's longhand counterparts are found.

**Comment by evilebottnawi** [comment]
@davidtheclark yep, i agree only throw errors on 4 properties, sorry, tired :sob:

**Comment by davidtheclark** [comment]
> sorry, tired 😭
Ha, no problem! I think that quality of shorthand properties stands out to me because it's bitten me so many times.

**Comment by evilebottnawi** [comment]
@davidtheclark we need name for rule :smile:

**Comment by davidtheclark** [comment]
Thinking about this more, I even wonder if there should be a `declaration-no-shorthand-property` rule. I'd use it.

**Comment by jeddy3** [comment]
> i agree only throw errors on 4 properties,

Phew, I'm glad we got there in the end :) @davidtheclark Thanks for helping to clarify!

> we need name for rule

`declaration-block-no-redundant-longhand-properties`?

- `declaration-block` because, like `declaration-block-no-duplicate-properties`, 
`declaration-block-no-ignored-properties` and
`declaration-block-no-shorthand-property-override`, it's *scoped to* declaration blocks. 
- `redundant` because that's aligned with `shorthand-property-no-redundant-values`.
- `longhand-properties`, well, because that's the thing that's redundant.

I guess all the shorthand properties ought to be caught by this rule (e.g. `font`, `border` etc), rather than just `margin` and `padding`.

> Thinking about this more, I even wonder if there should be a declaration-no-shorthand-property rule. I'd use it.

Ha, I know what you mean. The *implicit* setting of values by shorthand is a killer. Same goes for things like the `font` shorthand property`. Anyhow, lets create a separate issue for that discussion though :)

**Comment by evilebottnawi** [comment]
it is a good rule for the core
@jeddy3 sound good, i started today this rule, late we can rename
@davidtheclark any objectives?

**Comment by jeddy3** [comment]
- Name: `declaration-block-no-redundant-longhand-properties`
- Primary option: `true`
- Secondary options: none
- Message: `Unexpected longhand properties "${props}"`\* 
- Section: "Declaration block"

\*e.g. `Unexpected longhand properties "border-*-width"` if there is all four border width props i.e. `border-top-width`, `border-bottom-width` etc...

**Comment by evilebottnawi** [comment]
@jeddy3 for compat with `csslint` we should add option `ignoreProperties`, because `csslint` support this only for `padding` and `margin`

**Comment by ntwb** [comment]
Agreed, for whatever reasoning allowing an array of properties to be ignored should be included 👍
